File: Source/Libraries/GSF.Unmanaged/Collections/Types/HistorianValue.cs
//******************************************************************************************************
//  HistorianValue.cs - Gbtc
//
//  Copyright © 2013, Grid Protection Alliance.  All Rights Reserved.
//
//  Licensed to the Grid Protection Alliance (GPA) under one or more contributor license agreements. See
//  the NOTICE file distributed with this work for additional information regarding copyright ownership.
//  The GPA licenses this file to you under the Eclipse Public License -v 1.0 (the "License"); you may
//  not use this file except in compliance with the License. You may obtain a copy of the License at:
//
//      http://www.opensource.org/licenses/eclipse-1.0.php
//
//  Unless agreed to in writing, the subject software distributed under the License is distributed on an
//  "AS-IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. Refer to the
//  License for the specific language governing permissions and limitations.
//
//  Code Modification History:
//  ----------------------------------------------------------------------------------------------------
//  4/12/2013 - Steven E. Chisholm
//       Generated original version of source code. 
//     
//******************************************************************************************************


using System;
using System.Text;
using GSF;
using GSF.IO;

namespace openHistorian.Collections
{
    public struct HistorianValueStruct
    {
        public ulong Value1;
        public ulong Value2;
        public ulong Value3;

        public float AsSingle
        {
            get
            {
                return BitMath.ConvertToSingle(Value1);
            }
            set
            {
                Value1 = BitMath.ConvertToUInt64(value);
            }
        }

        public string AsString
        {
            get
            {
                byte[] data = new byte[16];
                BitConverter.GetBytes(Value1).CopyTo(data, 0);
                BitConverter.GetBytes(Value2).CopyTo(data, 8);
                return ASCIIEncoding.ASCII.GetString(data);
            }
            set
            {
                if (value.Length > 16)
                    throw new OverflowException("String cannot be larger than 16 characters");
                byte[] data = new byte[16];
                ASCIIEncoding.ASCII.GetBytes(value).CopyTo(data, 0);
                Value1 = BitConverter.ToUInt64(data, 0);
                Value2 = BitConverter.ToUInt64(data, 8);
            }
        }
    }

    public class HistorianValue
        : HistorianValueBase<HistorianValue>
    {
        /// <summary>
        /// Value 1 should be where the first 64 bits of the field is stored. For 32 bit values, use this field only.
        /// </summary>
        public ulong Value1;
        /// <summary>
        /// Should only be used if value cannot be entirely stored in Value1. Compression penality occurs when using this field.
        /// </summary>
        public ulong Value2;
        /// <summary>
        /// Should contain any kind of digital data such as Quality. Compression penality occurs when used for any other type of field.
        /// </summary>
        public ulong Value3;

        public override bool IsEqualTo(HistorianValue other)
        {
            return Value1 == other.Value1 && Value2 == other.Value2 && Value3 == other.Value3;
        }

        public override void CopyTo(HistorianValue other)
        {
            other.Value1 = Value1;
            other.Value2 = Value2;
            other.Value3 = Value3;
        }

        public override void Write(BinaryStreamBase stream)
        {
            stream.Write(Value1);
            stream.Write(Value2);
            stream.Write(Value3);
        }

        public override void Read(BinaryStreamBase stream)
        {
            Value1 = stream.ReadUInt64();
            Value2 = stream.ReadUInt64();
            Value3 = stream.ReadUInt64();
        }

        public override void WriteCompressed(BinaryStreamBase stream, HistorianValue previousValue)
        {
            stream.Write7Bit(previousValue.Value1 ^ Value1);
            stream.Write7Bit(previousValue.Value2 ^ Value2);
            stream.Write7Bit(previousValue.Value3 ^ Value3);
        }

        public override void ReadCompressed(BinaryStreamBase stream, HistorianValue previousValue)
        {
            Value1 = stream.Read7BitUInt64() ^ previousValue.Value1;
            Value2 = stream.Read7BitUInt64() ^ previousValue.Value2;
            Value3 = stream.Read7BitUInt64() ^ previousValue.Value3;
        }

        public override void Clear()
        {
            Value1 = 0;
            Value2 = 0;
            Value3 = 0;
        }

        public HistorianValue Clone()
        {
            HistorianValue value = new HistorianValue();
            value.Value1 = Value1;
            value.Value2 = Value2;
            value.Value3 = Value3;
            return value;
        }

        public HistorianValueStruct ToStruct()
        {
            return new HistorianValueStruct()
                {
                    Value1 = Value1,
                    Value2 = Value2,
                    Value3 = Value3
                };
        }

        public float AsSingle
        {
            get
            {
                return BitMath.ConvertToSingle(Value1);
            }
            set
            {
                Value1 = BitMath.ConvertToUInt64(value);
            }
        }

        public string AsString
        {
            get
            {
                byte[] data = new byte[16];
                BitConverter.GetBytes(Value1).CopyTo(data, 0);
                BitConverter.GetBytes(Value2).CopyTo(data, 8);
                return ASCIIEncoding.ASCII.GetString(data);
            }
            set
            {
                if (value.Length > 16)
                    throw new OverflowException("String cannot be larger than 16 characters");
                byte[] data = new byte[16];
                ASCIIEncoding.ASCII.GetBytes(value).CopyTo(data, 0);
                Value1 = BitConverter.ToUInt64(data, 0);
                Value2 = BitConverter.ToUInt64(data, 8);
            }
        }
    }
}

File: Source/Libraries/PythonAPIModule/openHistorian/historianValue.py
#******************************************************************************************************
#  historianValue.py - Gbtc
#
#  Copyright © 2021, Grid Protection Alliance.  All Rights Reserved.
#
#  Licensed to the Grid Protection Alliance (GPA) under one or more contributor license agreements. See
#  the NOTICE file distributed with this work for additional information regarding copyright ownership.
#  The GPA licenses this file to you under the MIT License (MIT), the "License"; you may not use this
#  file except in compliance with the License. You may obtain a copy of the License at:
#
#      http://opensource.org/licenses/MIT
#
#  Unless agreed to in writing, the subject software distributed under the License is distributed on an
#  "AS-IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. Refer to the
#  License for the specific language governing permissions and limitations.
#
#  Code Modification History:
#  ----------------------------------------------------------------------------------------------------
#  02/02/2021 - J. Ritchie Carroll
#       Generated original version of source code.
#
#******************************************************************************************************

from snapDB.snapTypeBase import snapTypeBase
from snapDB.enumerations import QualityFlags
from gsf.binaryStream import binaryStream
from gsf import Limits, ByteSize, override
from uuid import UUID
import struct
import numpy as np

class historianValue(snapTypeBase):
    """
    The standard SNAPdb value used for the openHistorian.
    """

    SnapTypeID = UUID("24dde7dc-67f9-42b6-a11b-e27c3e62d9ef")

    def __init__(self):
        # Value 1 should be where the first 64 bits of the field is stored. For 32 bit values, use this field only.
        self.Value1 = np.uint64(0)
        
        # Should only be used if value cannot be entirely stored in Value1. Compression penalty occurs when using this field.
        self.Value2 = np.uint64(0)
        
        # Should contain any kind of digital data such as Quality. Compression penalty occurs when used for any other type of field.
        self.Value3 = np.uint64(0)

    @property
    @override
    def TypeID(self) -> UUID:
        """
        Gets the Guid uniquely defining this SNAPdb type. 
        """        
        return historianValue.SnapTypeID

    @property
    @override
    def Size(self) -> int:
        """
        Gets the size of this SNAPdb type when serialized.
        """      
        return 24

    @override
    def SetMin(self):
        """
        Sets the provided SNAPdb type to its minimum value.
        """        
        self.Value1 = np.uint64(0)
        self.Value2 = np.uint64(0)
        self.Value3 = np.uint64(0)

    @override
    def SetMax(self):
        """
        Sets the provided SNAPdb type to its maximum value.
        """        
        self.Value1 = np.uint64(Limits.MAXUINT64)
        self.Value2 = np.uint64(Limits.MAXUINT64)
        self.Value3 = np.uint64(Limits.MAXUINT64)

    @override
    def Clear(self):
        """
        Clears the SNAPdb type.
        """        
        self.SetMin()

    @override
    def Read(self, stream: binaryStream):
        """
        Reads this SNAPdb type from the stream.
        """        
        self.Value1 = stream.ReadUInt64()
        self.Value2 = stream.ReadUInt64()
        self.Value3 = stream.ReadUInt64()
    
    @override
    def Write(self, stream: binaryStream):
        """
        Writes this SNAPdb type to the stream.
        """
        stream.WriteUInt64(self.Value1)
        stream.WriteUInt64(self.Value2)
        stream.WriteUInt64(self.Value3)

    @override
    def CopyTo(self, destination: "historianValue"):
        """
        Copies this SNAPdb type to the `destination`
        """        
        destination.Value1 = self.Value1;
        destination.Value2 = self.Value2;
        destination.Value3 = self.Value3;
    
    @override
    def CompareTo(self, other: "historianValue"):
        """
        Compares this SNAPdb type to the `other`
        """
        if self.Value1 < other.Value1:
            return -1
        if self.Value1 > other.Value1:
            return 1
        if self.Value2 < other.Value2:
            return -1
        if self.Value2 > other.Value2:
            return 1
        if self.Value3 < other.Value3:
            return -1
        if self.Value3 > other.Value3:
            return 1
        return 0
    
    @property
    def AsSingle(self) -> np.float32:
        """
        Gets `Value1` type cast as a single.
        """
        return np.frombuffer(int(np.uint32(self.Value1)).to_bytes(ByteSize.UINT32, "little"), np.float32)[0]

    @AsSingle.setter
    def AsSingle(self, value: np.float32):
        """
        Sets `Value1` type cast from a single.
        """
        buffer = bytearray(8)
        valueBytes = struct.pack("<f", value)

        for i in range(4):
            buffer[i] = valueBytes[i]

        self.Value1 = np.frombuffer(buffer, np.uint64)[0]

    @property
    def AsQuality(self) -> QualityFlags:
        """
        Gets `Value3` type cast as `QualityFlags`.
        """
        return QualityFlags(self.Value3)

    @AsQuality.setter
    def AsQuality(self, value: QualityFlags):
        """
        Sets `Value3` type cast from `QualityFlags`.
        """
        self.Value3 = np.uint64(value.value)

    def ToString(self) -> str:
        return f"{self.AsSingle:.3f} [{self.AsQuality}]"